lib/elixir/lib/kernel/struct.ex

```diff
@@ -0,0 +1,193 @@
+defmodule Kernel.Struct do
+  @moduledoc false
+
+  @doc false
+  def expand(struct_or_match, {:%{}, map_meta, map_args}, %{context: context} = env) do
+    meta = [file: env.file, line: env.line]
+    map_args = clean_struct_key_from_map_args(meta, map_args, env)
+
+    {[struct_or_match, map], env} =
+      :elixir_expand.expand_args([struct_or_match, {:%{}, map_meta, map_args}], env)
```

Comment:
We can't really call this here. Because macros cannot return the updated environment, if you expand anything that changes the environment here, such as an import or alias or require, it will be lost.

We can only really expand the first argument (the struct name) and by using Macro.expand. But you have to be careful with bootstrapping. So what I would do is to expand the first argument on kernel before calling this module.

Then the keys have to be expanded manually to with Macro.expand/2.

Reply:
Thanks!